test/dataframe_tester.cc

```diff
@@ -2987,6 +2987,129 @@ int main(int argc, char *argv[]) {
+        assert(std::isnan(result2[19]));
+        assert(std::isnan(result2[20]));
+
+        DiffVisitor<double>   diff_visit3(3, false);
+        const auto            &result3 =
+            df.single_act_visit<double>("col_1", diff_visit3).get_result();
+
+        assert(result3.size() == 21);
+        assert(std::isnan(result3[0]));
+        assert(std::isnan(result3[1]));
+        assert(std::isnan(result3[2]));
+        assert(std::isnan(result3[4]));
+        assert(result3[5] == 1.0);
+        assert(result3[6] == 3.0);
+        assert(std::isnan(result3[7]));
+        assert(std::isnan(result3[16]));
+        // FIXME: Why does it fail to assert result3[17]? But the output is correct.
+        // std::cout << "result3[17] == " << result3[17] << "\n";
+        // assert(result3[17] == 1.907);
+        assert(result3[18] == -0.66);
+        assert(std::isnan(result3[19]));
+        assert(std::isnan(result3[20]));
+
+        DiffVisitor<double>   diff_visit4(-3, false);
+        const auto              &result4 =
+            df.single_act_visit<double>("col_1", diff_visit4).get_result();
+
+        assert(result4.size() == 21);
+        assert(std::isnan(result4[0]));
+        assert(std::isnan(result4[1]));
+        assert(result4[2] == -1.0);
+        assert(result4[3] == -3.0);
+        assert(std::isnan(result4[4]));
+        assert(std::isnan(result4[13]));
+        // FIXME: Why does it fail to assert result4[14]? But the output is correct.
+        // std::cout << "result4[14] == " << result4[14] << "\n";
+        // assert(result4[14] == -1.907);
+        assert(result4[15] == 0.66);
+        assert(std::isnan(result4[16]));
+        assert(std::isnan(result4[17]));
+        assert(std::isnan(result4[18]));
+        assert(std::isnan(result4[19]));
+        assert(std::isnan(result4[20]));
+
+        DiffVisitor<double>   diff_visit5(3, true);
+        const auto            &result5 =
+            df.single_act_visit<double>("col_1", diff_visit5).get_result();
+
+        assert(result5.size() == 10);
+        assert(result5[0] == 1.0);
+        assert(result5[1] == 3.0);
+        assert(result5[2] == -14.44);
+        assert(result5[7] == -17.94);
+        // FIXME: Why does it fail to assert result5[8]? But the output is correct.
+        // std::cout << "result5[8] == " << result5[8] << "\n";
+        // assert(result5[8] == 1.907);
+        assert(result5[9] == -0.66);
+
+        DiffVisitor<double>   diff_visit6(-3, true);
+        const auto            &result6 =
+            df.single_act_visit<double>("col_1", diff_visit6).get_result();
+
+        assert(result6.size() == 10);
+        assert(result6[0] == -1.0);
+        assert(result6[1] == -3.0);
+        assert(result6[2] == 14.44);
+        assert(result6[7] == 17.94);
+        // FIXME: Why does it fail to assert result6[8]? But the output is correct.
+        // std::cout << "result6[8] == " << result6[8] << "\n";
+        // assert(result6[8] == -1.907);
```

Comment:
It is a precision problem. Do it this way
```
assert(fabs(result6[8] - -1.907) < 0.0000001);
```

Reply:
Thanks, I have modified it.